identity_stardust/src/did/stardust_did.rs

```diff
@@ -0,0 +1,927 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use core::convert::TryFrom;
+use core::fmt::Debug;
+use core::fmt::Display;
+use core::fmt::Formatter;
+use core::str::FromStr;
+
+use identity_core::common::KeyComparable;
+use iota_client::block::output::AliasId;
+use serde::Deserialize;
+use serde::Serialize;
+
+use identity_did::did::BaseDIDUrl;
+use identity_did::did::CoreDID;
+use identity_did::did::DIDError;
+use identity_did::did::DIDUrl;
+use identity_did::did::DID;
+
+use crate::NetworkName;
+
+pub type Result<T> = std::result::Result<T, DIDError>;
+
+/// A DID URL conforming to the IOTA Stardust DID method specification.
+///
+/// See [`DIDUrl`].
+pub type StardustDIDUrl = DIDUrl<StardustDID>;
+
+// The length of an AliasID, which is a BLAKE2b-256 hash (32-bytes).
+const TAG_BYTES_LEN: usize = 32;
+
+/// A DID conforming to the IOTA UTXO DID method specification.
+///
+/// This is a thin wrapper around the [`DID`][`CoreDID`] type from the
+/// [`identity_did`][`identity_did`] crate.
+#[derive(Clone, Hash, Debug, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
+#[repr(transparent)]
+#[serde(into = "CoreDID", try_from = "CoreDID")]
+pub struct StardustDID(CoreDID);
+
+impl StardustDID {
+  /// The URL scheme for Decentralized Identifiers.
+  pub const SCHEME: &'static str = CoreDID::SCHEME;
+
+  /// The IOTA UTXO DID method name (`"stardust"`).
+  // TODO: This will be changed to `iota` in the future.
+  pub const METHOD: &'static str = "stardust";
+
+  /// The default Tangle network (`"main"`).
+  // TODO: Change to use `NetworkName` once that has been ported.
+  pub const DEFAULT_NETWORK: NetworkName = NetworkName::main();
+
+  /// Converts an owned [`CoreDID`] to a [`StardustDID`].
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the input does not conform to the [`StardustDID`] specification.
+  pub fn try_from_core(did: CoreDID) -> Result<Self> {
+    Self::check_validity(&did)?;
+
+    Ok(Self(Self::normalize(did)))
+  }
+
+  /// Constructs a new [`StardustDID`] from a byte representation of the tag and the given network name.
+  pub fn new<T: Into<NetworkName>>(bytes: &[u8; 32], network_name: T) -> Self {
```

Comment:
Aside: a doc-test might be nice to show what it looks like. Optional.

```suggestion
  /// Constructs a new [`StardustDID`] from a byte representation of the tag and the given network name.
  ///
  /// See also [`StardustDID::placeholder`].
  ///
  /// # Example
  ///
  /// ```
  /// # use identity_did::did::DID;
  /// # use identity_stardust::NetworkName;
  /// # use identity_stardust::StardustDID;
  /// #
  /// let placeholder = StardustDID::new(&[1;32], NetworkName::try_from("smr").unwrap());
  /// assert_eq!(placeholder.as_str(), "did:stardust:smr:0x0101010101010101010101010101010101010101010101010101010101010101");
  /// ```
  pub fn new<T: Into<NetworkName>>(bytes: &[u8; 32], network_name: T) -> Self {
```